### PR TITLE
feature/CDAP-20322: implement push app task and RemoteSourceControlOperationRunner

### PR DESCRIPTION
**Pull request by de-lan**
## What

1. Implemented RemoteSourceControlOperationRunner#push
2. Conditional SourceControlOperationRunner guice binding on `Constants.TaskWorker.POOL_ENABLE`

### CODE REVIEW

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java [hunk 1]
```diff
@@ -14,21 +14,33 @@
...
+  private final CConfiguration cConf;
+
+  public SourceControlModule(CConfiguration cConf) {
+    this.cConf = cConf;
+  }
   @Override
   protected void configure() {
-    // TODO: CDAP-20322 add the remoteSourceControlOperationRunner
-    bind(SourceControlOperationRunner.class).to(InMemorySourceControlOperationRunner.class).in(Scopes.SINGLETON);
+    boolean workerPoolEnabled = cConf.getBoolean(Constants.TaskWorker.POOL_ENABLE);
```
Comment (de-lan) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: Should we use the factory/provider pattern or this conditional binding is sufficient? @chtyim @albertshau

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/SourceControlOperationRunner.java [hunk 2]
```diff
@@ -17,27 +17,24 @@
...
+   * @param pushAppContext {@link PushAppContext} pf the application to be pushed
+   * @return {@link PushAppResponse} file-paths and file-hashes for the updated configs.
+   * @throws PushFailureException when the push operation fails for any unexpected reason.
+   * @throws NoChangesToPushException when the there's no change for the application to push.
+   * @throws AuthenticationConfigException when the repository configuration is invalid.
    */
-  PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
-                       ApplicationDetail appToPush, CommitMeta commitDetails)
-    throws PushFailureException, NoChangesToPushException, AuthenticationConfigException;
+  PushAppResponse push(PushAppContext pushAppContext) throws PushFailureException, NoChangesToPushException,
```
Comment (de-lan) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/SourceControlOperationRunner.java:36]: Use a context class for better consistency because we need to serialize the params to `RunnableTaskRequest` for remote implementation

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java [hunk 3]
```diff
@@ -14,21 +14,33 @@
...
+    if (workerPoolEnabled) {
+      bind(SourceControlOperationRunner.class).to(RemoteSourceControlOperationRunner.class).in(Scopes.SINGLETON);
+    } else {
+      bind(SourceControlOperationRunner.class).to(InMemorySourceControlOperationRunner.class).in(Scopes.SINGLETON);
+    }
```
Comment (arjan-bal) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: Just curious, why do we make these singletons?

Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: Actually I don't have a strong opinion if they should be singletons..

Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: removed singleton scope

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import io.cdap.cdap.security.auth.context.AuthenticationContextModules;
+import io.cdap.cdap.security.guice.SecureStoreClientModule;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+/**
+ * The utility class that assists source control management tasks.
+ */
+public class TaskUtil {
+  public static Injector createSourceControlTaskInjector(CConfiguration cConf,
```
Comment (de-lan) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java:35]: Having this static method out to be reused by upcoming `PullAppTask` and `ListAppsTask`, since they all share the same injector for `SourceControlOperationRunner`

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import io.cdap.cdap.security.auth.context.AuthenticationContextModules;
+import io.cdap.cdap.security.guice.SecureStoreClientModule;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+/**
+ * The utility class that assists source control management tasks.
+ */
+public class TaskUtil {
+  static Injector createSourceControlTaskInjector(CConfiguration cConf,
```
Comment (de-lan) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java]: Having this package private static method out to be reused by upcoming `PullAppTask` and `ListAppsTask`, since they all share the same injector for `SourceControlOperationRunner`

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java (hunk 1 above)
Comment (tivv) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: I would prefer a provider method. We generally try not to add cConf to the module constructor, but inject it

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    } catch (Exception ex) {
+      throw new PushFailureException(ex.getMessage(), ex.getCause());
+    }
+  }
+
+  @Override
+  public PullAppResponse<?> pull(ApplicationReference appRef, RepositoryConfig repoConfig)
+    throws PullFailureException, NotFoundException, AuthenticationConfigException {
+    // TODO: CDAP-20356, pull application in task worker
+    return null;
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java]: It's better to throw an `UnsupportedOperationException` as it allows to fail early

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        throw new NoChangesToPushException(exceptionMessage);
+      }
+
+      if (AuthenticationConfigException.class.getName().equals(exceptionClass)) {
+        throw new AuthenticationConfigException(exceptionMessage, remoteTaskException.getCause());
+      }
+
+      throw new PushFailureException(exceptionMessage, remoteTaskException.getCause());
+    } catch (Exception ex) {
+      throw new PushFailureException(ex.getMessage(), ex.getCause());
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java]: Why are you dropping data from the `ex`? It may have important stack trace and no cause (e.g. if it's NPE)

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      
+      LOG.debug("Pushing application {} to linked repository", pushAppContext.getAppToPush());
+      byte[] result = remoteTaskExecutor.runTask(request);
+      return GSON.fromJson(new String(result, StandardCharsets.UTF_8), PushAppResponse.class);
+    } catch (RemoteExecutionException e) {
+      RemoteTaskException remoteTaskException = e.getCause();
+      String exceptionClass = remoteTaskException.getRemoteExceptionClassName();
+      String exceptionMessage = remoteTaskException.getMessage();
+
+      if (NoChangesToPushException.class.getName().equals(exceptionClass)) {
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: I am wondering if we should just improve io.cdap.cdap.api.service.worker.RemoteExecutionException#fromBasicThrowable to instantiate the remove class. The way we do here looses any root causes

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import io.cdap.cdap.common.internal.remote.RunnableTaskModule;
+import io.cdap.cdap.security.auth.context.AuthenticationContextModules;
+import io.cdap.cdap.security.guice.SecureStoreClientModule;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+/**
+ * The utility class that assists source control management tasks.
+ */
+public class TaskUtil {
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java]: Please follow utility class pattern (final, private constructor)

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    discoveryService.register(URIScheme.createDiscoverable(Constants.Service.SECURE_STORE_SERVICE, httpService));
+
+    // Set up the mock objects
+    mockInMemoryOperationRunner = Mockito.mock(InMemorySourceControlOperationRunner.class);
+  }
+  
+  @Before
+  public void resetMock() throws Exception {
+    Mockito.reset(mockInMemoryOperationRunner);
+    PowerMockito.spy(TaskUtil.class);
```
Comment (tivv) [inline cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: Static method mocking is a code smell. We should avoid it on new code

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    this.cConf = cConf;
+    this.discoveryService = discoveryService;
+    this.discoveryServiceClient = discoveryServiceClient;
+  }
+
+  @Override
+  public void run(RunnableTaskContext context)
+    throws AuthenticationConfigException, PushFailureException, NoChangesToPushException, IOException {
+    PushAppContext pushContext = GSON.fromJson(context.getParam(), PushAppContext.class);
+    SourceControlOperationRunner operationRunner =
```
Comment (tivv) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: Could you make this method non-static? It would help a lot with proper structure / testing. Basically, it would become an injector factory

BTW: Why do we need to create an injector on the spot?

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java [hunk 12]
```diff
@@ -14,21 +14,32 @@
...
 /**
  *  Guice module for source control management classes.
  */
 public class SourceControlModule extends AbstractModule {
 
+  private final CConfiguration cConf;
+
+  public SourceControlModule(CConfiguration cConf) {
+    this.cConf = cConf;
+  }
```
Comment (albertshau) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: nit: newline between methods

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java (hunk 1 above)
Comment (albertshau) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: +1 I think that's the existing pattern too.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 8 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: Actually the root causes are propagated in `RemoteTaskException`, we can get it from `RemoteTaskException#getCause`. I added the cause constructor for `NoChangesToPushException`, it should be propagating the statckTrace now.

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 7 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java]: ooops, my mistake. Added back

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 6 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java]: Throw `UnsupportedOperationException` now

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * The task that pushes an application to linked repository.
+ */
+public class PushAppTask implements RunnableTask {
+
+  private static final Gson GSON = new GsonBuilder().create();
```
Comment (albertshau) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: nit: simpler to do `new Gson()`, unless you need to do some customization of the gson

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java (hunk 11 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: So currently for TASK_WORKER guice injection, only `CConfiguration, DiscoveryService, DiscoveryServiceClient` are injected, [referring to](https://github.com/cdapio/cdap/blob/develop/cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/TaskWorkerHttpHandlerInternal.java#L100) `io.cdap.cdap.common.internal.remote.RunnableTaskLauncher`. Currently TASK_WORKER `RunnableTask`s implementations manage the injection itself at runtime, the examples are `ConfiguratorTask`, `SystemAppTask`... These TASK_WORKER tasks all uses a static injection creator that takes in the `DiscoveryService` as well as other necessary components like the `RemoteSecureStore`, etc

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java (hunk 9 above)
Comment (albertshau) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java]: try to avoid having util classes if possible, they tend to accumulate unrelated logic over time and encourage poor class design.

In this particular case, maybe have an abstract SourceControlTask class that creates the InMemorySourceControlOperationRunner?

Or make this into a guice module class instead?

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java (hunk 12 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: added newline

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java (hunk 13 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: nice suggestion, updated

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java (hunk 9 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java]: Will try the abstract class approach. But I think the biggest blocker is how to mock in the tests

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java (hunk 11 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: removed the static method now

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java (hunk 10 above)
Comment (de-lan) [reply cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: removed, now the tests will run the actual tasks again a test bare repo

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java (hunk 9 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/TaskUtil.java]: Updated to the SourceControlTask class approach, pls have a second round of review, thanks! @tivv @albertshau

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java (hunk 1 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java]: Updated to Provider pattern

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+              DiscoveryServiceClient discoveryServiceClient) {
+    super(cConf, discoveryService, discoveryServiceClient);
+  }
+
+  @Override
+  public void run(RunnableTaskContext context)
+    throws AuthenticationConfigException, PushFailureException, NoChangesToPushException, IOException {
+    PushAppContext pushContext = GSON.fromJson(context.getParam(), PushAppContext.class);
+
+    LOG.debug("Pushing application {} in worker.", pushContext.getAppToPush().getName());
```
Comment (samdgupi) [inline cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: nit: Should it be Info log? debug level might not enabled and this logline will be helpful for us for debugging

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java [hunk 15]
```diff
@@ -161,7 +161,7 @@ protected void before() throws Exception {
   }
 
   @Override
-  protected void after() {
+  public void after() {
```
Comment (samdgupi) [inline cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java:164]: why was this changed ?

Comment (de-lan) [reply cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java:164]: It was used in the test to shutdown remote repo

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java (hunk 14 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/worker/PushAppTask.java]: nice catch, changed to `Info`

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 8 above)
Comment (albertshau) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: I think the problem here is that the exception classes in the cause are not the same because of how RemoteExecutionException is implemented, plus everything that calls runTask() needs to do similar logic to propagate the exception. +1 to improving it, though it should be in a separate PR. Can you open an enhancement jira and add a todo?

Comment (albertshau) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: what about the other exceptions like AuthenticationConfigException, PushFailureException, and IOException?

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+      protected void configure() {
+        bind(RemoteClientFactory.class).toInstance(remoteClientFactory);
+      }
+    }).getInstance(RemoteSecureStore.class);
+    remoteSecureStore.put(NAMESPACE1, TOKEN_NAME, MOCK_TOKEN, "fake git token", ImmutableMap.of("prop1", "value1"));
+  }
+
+  @Before
+  public void beforeEach() throws Exception {
+    cConf.setInt(Constants.SourceControlManagement.GIT_COMMAND_TIMEOUT_SECONDS, GIT_COMMAND_TIMEOUT);
```
Comment (albertshau) [inline cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: seems like this can be done once in the `@BeforeClass` method?

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+    cConf.set(Constants.SourceControlManagement.GIT_REPOSITORIES_CLONE_DIRECTORY_PATH,
```
Comment (albertshau) [inline cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: why do we need this, I thought each clone happens in it's own directory within this base directory?

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 8 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: Added TODO ticket: https://cdap.atlassian.net/browse/CDAP-20418

`AuthenticationConfigException` is handled in below if section. `PushFailureException`, and `IOException` are both transformed to `PushFailureException`

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java (hunk 16 above)
Comment (de-lan) [reply cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: moved up

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java (hunk 17 above)
Comment (de-lan) [reply cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: You're correct, each clone happens in it's own directory. Just that in test, we wanted it to clone in a different root directory in the test `/target`, instead of the default one as `/tmp/source-control/` in cdap-default.xml

Comment (de-lan) [reply cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunnerTest.java]: Also moved to `@BeforeClass`

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 8 above)
Comment (albertshau) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: shouldn't it do the same thing and try to preserve the IOException? Otherwise the remote and local implementations will throw different exceptions in different situations.

If there isn't really a meaningful way to handle IOException, we should wrap it into a PushFailureException in the other implementation as well so that they are consistent.

**Review by albertshau — APPROVED** [review]

File: cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java (hunk 8 above)
Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: @albertshau We don't throw `IOException` in the local implementation either (it was wrapped into a `PushFailureException`. This `IOException` in remote implementation was thrown by the `PushAppTask#context.writeResult`.

Comment (de-lan) [reply cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java:80]: Offline discusses. IOException is no need to preserve
